### PR TITLE
RS/YJ/Rule 5-1

### PR DESCRIPTION
**Pull request by yunjoonjung-PNNL**
This Rule hasn't been tested yet. I'll work with Jackson to make this rule pass. Thanks!

### CODE REVIEW

**Comment by JacksonJ-KC** [comment]
@yunjoonjung-PNNL This draft was helpful. I ran the tests with the built-in validation and found a couple issues with the previous test json that I did not catch with the local validation that I was running previously.

This test is not in the TCD spreadsheet because it requires multiple RMDs and the spreadsheet method would have added complications. This json was created by copying a snippet of another test json and altering manually.

**Comment by yunjoonjung-PNNL** [comment]
> @yunjoonjung-PNNL This draft was helpful. I ran the tests with the built-in validation and found a couple issues with the previous test json that I did not catch with the local validation that I was running previously.
> 
> This test is not in the TCD spreadsheet because it requires multiple RMDs and the spreadsheet method would have added complications. This json was created by copying a snippet of another test json and altering manually.

Thank you for your quick response and No worries! Please refer to this PR for 5-1 TCD creation and let me know if this rule has any issues.

**Comment by JacksonJ-KC** [comment]
See [https://github.com/pnnl/ruleset-checking-tool/pull/1171](https://github.com/pnnl/ruleset-checking-tool/pull/1171)

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+                    (
+                        key
+                        for key, range_ in azimuth_fen_area_dict_b.items()
+                        if int(key.split("-")[0]) <= azi <= int(key.split("-")[1])
+                    ),
+                    None,
+                )
+
+                azimuth_fen_area_dict_b = {
+                    f"{azi}-{azi+3}": ZERO.AREA for azi in range(3, 360, 3)
```
Comment (JacksonJ-KC) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:152]: It looks like `azimuth_fen_area_dict_b` is being prepopulated with strings beginning with "3-6". Should the range "0-3" get included?

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                }
+                total_surface_fenestration_area_b = ZERO.AREA
+                for surface_b in find_all(
+                    "$.building_segments[*].zones[*].surfaces[*]", building_b0
+                ):
+                    if get_opaque_surface_type(surface_b) == OST.ABOVE_GRADE_WALL:
+                        surface_azimuth_b = surface_b["azimuth"]
+
+                        if surface_azimuth_b not in azimuth_fen_area_dict_b:
```
Comment (JacksonJ-KC) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:161]: surface_azimuth_b I believe is an integer and will not be `in azimuth_fen_area_dict_b` since it was prepopulated with strings

Comment (JacksonJ-KC) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:161]: I was thinking that it could be good to populate the dictionary with integers in range(3, 360, 3), then round `surface_azimuth_b` to the nearest 3 degrees

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+                has_basseline_270_output = has_baseline_270 and find_one(
+                    "$.output.output_instance", building_b270, False
+                )
+
+                # define a function to find the azimuth's corresponding key
+                find_key_for_azi = lambda azi: next(
+                    (
+                        key
+                        for key, range_ in azimuth_fen_area_dict_b.items()
+                        if int(key.split("-")[0]) <= azi <= int(key.split("-")[1])
```
Comment (JacksonJ-KC) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:146]: I think azi is a pint object here and cannot be compared with pure integers. 

this helped to get past the error:
if int(key.split("-")[0]) <= azi.to('degrees').magnitude <= int(key.split("-")[1])

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 1 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:152]: Good point! Addressed.

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 2 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:161]: Good catch. These lines are not needed because ```ZERO.AREA``` was already set up above as a default value.

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 3 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:146]: Thanks Jackson, Addressed.

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+                has_proposed = data["has_proposed"]
+                has_user = data["has_user"]
+                baseline_list = data["baseline_list"]
+                no_of_rmds = data["no_of_rmds"]
+                no_of_output_instance = data["no_of_output_instance"]
+
+                has_baseline_0 = RULESET_MODEL.BASELINE_0 in baseline_list
+                has_baseline_90 = RULESET_MODEL.BASELINE_90 in baseline_list
+                has_baseline_180 = RULESET_MODEL.BASELINE_180 in baseline_list
+                has_baseline_270 = RULESET_MODEL.BASELINE_270 in baseline_list
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:141]: I kind of feel the baseline_list is not necessary because context.BASELINE_X makes sure the context is the correct type.
You can do this without even use the baseline_list by simply checking whether `building_bx` is None or an Object.
Similarly, you do not need `has_proposed` or `has_user` from higher level because ``building_p` and `building_u` are from these two models. If they are None, then it means there are no proposed or user.

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+            no_of_output_instance = len(baseline_list)
+
+            return {
+                "has_proposed": has_proposed,
+                "has_user": has_user,
+                "baseline_list": baseline_list,
+                "no_of_rmds": no_of_rmds,
+                "no_of_output_instance": no_of_output_instance,
+            }
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:105]: After review this rule, I feel that you do not need this `create_data` function. All info can be retrieved inside `BuildingRule`

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+                )
+
+                # define a function to find the azimuth's corresponding key
+                find_key_for_azi = lambda azi: next(
+                    (
+                        key
+                        for key, _ in azimuth_fen_area_dict_b.items()
+                        if int(key.split("-")[0])
+                        <= azi.to(ureg("degree")).m
+                        <= int(key.split("-")[1])
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:169]: Will need to update this to `<`

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+                    ),
+                    None,
+                )
+
+                azimuth_fen_area_dict_b = {
+                    f"{azi}-{azi+3}": ZERO.AREA for azi in range(0, 360, 3)
+                }
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:176]: May need to update this to follow Jackson's suggestion (not prepopulate bins)

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+                    (
+                        rotation_expected_b
+                        and has_baseline_90
+                        and has_baseline_180
+                        and has_baseline_270
+                        and has_basseline_90_output
+                        and has_basseline_180_output
+                        and has_basseline_270_output
+                        and no_of_rmds == 5
+                        and no_of_output_instance == 6
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:277]: This is reversed, `no_of_rmds == 6 and no_of_output_instance == 5`

**Review by weilixu — COMMENTED** [review]
Some comments need to address

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 6 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:169]: Addressed

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 8 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:277]: Addressed

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+        def create_data(self, context, data):
+            rmd_b0 = context.BASELINE_0
+            rmd_b90 = context.BASELINE_90
+            rmd_b180 = context.BASELINE_180
+            rmd_b270 = context.BASELINE_270
+            rmd_p = context.PROPOSED
+            rmd_u = context.USER
+
+            has_proposed = True if find_one("$.type", rmd_p) else False
+            has_user = True if find_one("$.type", rmd_u, False) else False
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:77]: Good point. Addressed.

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 4 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:141]: Addressed

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 7 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:176]: Addressed.

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 10]
```diff
@@ -112,7 +112,7 @@ def create_data(self, context, data):
 
             # filter out baseline rmds that aren't None then count the length
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:113]: The comment does not seem the be correct. This includes baseline and proposed.

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 11]
```diff
@@ -112,7 +112,7 @@ def create_data(self, context, data):
...
             no_of_output_instance = len(
-                [rmd_b["type"] for rmd_b in baseline_rmd_list if rmd_b is not None]
+                [rmd["type"] for rmd in rmd_list_no_user if rmd is not None]
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:115]: Shouldn't this count the number of `output_instances` in the rmd files?
```python
len(compact([find_one($.output.output_instance) for rmd in rmd_list_no_user]))
```

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 12]
```diff
@@ -112,7 +112,7 @@ def create_data(self, context, data):
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:112]: For above code, a simple way to do is using the compact function from pydash.
```python
no_of_rmds = len(compact(rmd_list))
```

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 10 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:113]: Addressed.

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 11 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:115]: Great. Thank you Weili!

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 12 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:112]: Great point. Addressed.

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+
+                # define a function to get the azimuth's corresponding key
+                def get_key_for_azi(azi):
+                    azi_value = azi.to("degrees").m
+                    low_bound = azi_value - (azi_value % 3)
+                    high_bound = low_bound + 3
+                    return f"{low_bound}-{high_bound}"
+
+                azimuth_fen_area_dict_b = {}
+                total_surface_fenestration_area_b = ZERO.AREA
```
Comment (JacksonJ-KC) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:178]: ` total_surface_fenestration_area_b` needs to be reset to 0 in each iteration of surface

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+            )
+
+            return {
+                "has_baseline_0": has_baseline_0,
+                "has_baseline_90": has_baseline_90,
+                "has_baseline_180": has_baseline_180,
+                "has_baseline_270": has_baseline_270,
+                "has_proposed": has_proposed,
+                "has_user": has_user,
+                "has_basseline_0_output": has_basseline_0_output,
```
Comment (JacksonJ-KC) [inline rct229/rulesets/ashrae9012019/section5/section5rule1.py:125]: select all occurrences of basseline typos and replace with baseline

Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py:125]: Good catch. Addressed

File: rct229/rulesets/ashrae9012019/section5/section5rule1.py (hunk 13 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section5/section5rule1.py]: Addressed!

**Review by JacksonJ-KC — APPROVED** [review]
Looks good to me! Not sure if you need my approval, but you have it!

**Comment by yunjoonjung-PNNL** [comment]
> Looks good to me! Not sure if you need my approval, but you have it!

Thanks for your hard work throughout this rule!

File: rct229/rule_engine/rule_base.py [hunk 15]
```diff
@@ -286,7 +286,9 @@ def get_context(self, rmds, data=None):
             if self.rmrs_used[ruleset_model] and (
                 rmds[ruleset_model] is None or not rmds[ruleset_model]
             ):
-                missing_contexts.append(ruleset_model)
+                # BASELINE 90, 180, and 270 are optional
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rule_engine/rule_base.py:289]: revert this change

File: rct229/rule_engine/engine.py [hunk 16]
```diff
@@ -146,7 +146,7 @@ def evaluate_rules(
     rmds_used = get_rmd_instance()
     for rule in rules_list:
         for rule_model in rmds.get_ruleset_model_types():
-            if rule.rmrs_used[rule_model]:
+            if rule.rmrs_used[rule_model] and rmds[rule_model] is not None:
```
Comment (yunjoonjung-PNNL) [inline rct229/rule_engine/engine.py:149]: revert this change

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section5/rule_5_31.json [hunk 17]
```diff
@@ -82,7 +82,7 @@
                                 ]
                             }
                         ],
-                        "type": "USER"
+                        "type": "BASELINE_0"
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section5/rule_5_31.json:85]: @JasonGlazer @JacksonJ-KC Could you one of you reflect this change in the section 5 spreadsheet? I believe when section 5 was renumbered, this part was changed unexpectedly somehow.

**Comment by yunjoonjung-PNNL** [comment]
@weilixu This rule is ready to be reviewed again. I added Jackson's TCD from PR #1171 to make the CI/CD pass.

**Review by weilixu — APPROVED** [review]

**Comment by weilixu** [comment]
Approved but wait for the json test updates.

**Comment by JacksonJ-KC** [comment]
> Approved but wait for the json test updates.

Are you waiting on me for anything?

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section5/rule_5_31.json (hunk 17 above)
Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section5/rule_5_31.json:85]: This test json for rule 5-31 does not align with the section 5 spreadsheet - the `test_description`s belong to a different rule. The spreadsheet appears to be correct.

**Comment by yunjoonjung-PNNL** [comment]
@weilixu Jackson's work was merged into this branch. I believe this PR is ready to be merged into the ```develop``` branch. Thanks!
